### PR TITLE
Add support for exporting ssg hooks

### PR DESCRIPTION
**Pull request by zephraph**
This is the reapplication of #37, but with downstream fixes for the mdx error that was causing issues. 

The test here is failing for an odd reason which isn't at all related to this plugin. It's another upstream issue: https://github.com/mdx-js/mdx/pull/912. We actually don't have to wait for the upstream fix though, we could just rename the test. 

After that's not failing, the test will still fail for a reason I haven't determined yet. For whatever reason it just doesn't generate the output html file. I don't know if that's an issue with next, mdx, or this plugin. I'm leaning to one of the first two. I'm going to take a break from this for a few days, but feel free to poke at it.

I'll get all these issues sorted eventually 😅

### CODE REVIEW

**Comment by zephraph** [comment]
So the first upstream issue is fixed. Once that's published I'll merge the update into here and we can move onto looking at the next test failure.

**Comment by zephraph** [comment]
@jescalan are tests disabled for forks? I'm just now noticing they haven't ran for any of my PRs.

**Comment by jescalan** [comment]
Yeah I noticed this as well I'm not sure - let me look into this

**Comment by zephraph** [comment]
Hey @jescalan will you run these tests locally and see if everything is passing for you? If so, this PR is good to go.

**Comment by jescalan** [comment]
Sorry for the huge delay here @zephraph - just ran this locally and your newly added test is failing for me, even after a full `node_modules` clear and reinstall:

```
 (failed) exports ssg functions from layout

    ENOENT: no such file or directory, open 'xxx/next-mdx-enhanced/__tests__/fixtures/layout-exports-ssg/out/docs/intro.html'
```

**Comment by zephraph** [comment]
Okay, thanks. That was a similar failure as what I was seeing before. 

I’ll have to dig back through it to see what’s up. Does anything jump out at you? I think there’s just something wrong with how I set up the test.

**Comment by jescalan** [comment]
Didn't see anything obvious, but if you feel stuck I can absolutely dig deeper here - let me know!

**Comment by zephraph** [comment]
Hey @jescalan, I'm coming back to this after being away for a while...

My last commit only runs the one failing test and comments out the cleanup. The first time you run it the test fails (there's not exported static html). The second time you run it it succeeds. 

I'm somewhat struggling to understand if this is actually just a behavior of next (where it doesn't _necessarily_ always generate a static file on the first run even if it says it does) or if it's something else that I'm missing. 

Thoughts?

**Comment by zephraph** [comment]
Here are the results I get when running with cleanup disabled. The first run is considered "clean" (i.e. no artifacts from old builds), the second has artifacts from the first build. 

First run (failure)

```
 FAIL  __tests__/integration.js (6.323s)
  ● Console

    console.log node_modules/next/dist/build/spinner.js:3
      Creating an optimized production build...
    console.log node_modules/next/dist/build/index.js:10
      undefined
    console.log node_modules/next/dist/build/index.js:12
      Compiled successfully.
      
    console.log node_modules/next/dist/build/spinner.js:3
      Automatically optimizing pages...
    console.log node_modules/next/dist/build/index.js:44
      undefined
    console.log node_modules/next/dist/build/utils.js:7
      Page                                                           Size     First Load
      ─ ● /docs/intro                                                1.49 kB     54.1 kB
      + shared by all                                                52.6 kB
        ├ static/pages/_app.js                                       597 B
        ├ chunks/833cecfbdbc5190f1097354e088c9186d4c7a478.d90c6a.js  6.6 kB
        ├ chunks/framework.7dc243.js                                 40 kB
        ├ runtime/main.dff265.js                                     4.71 kB
        └ runtime/webpack.b65cab.js                                  746 B
    console.log node_modules/next/dist/build/utils.js:7
      undefined
    console.log node_modules/next/dist/build/utils.js:7
      λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
      ○  (Static)  automatically rendered as static HTML (uses no initial props)
      ●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
    console.log node_modules/next/dist/build/utils.js:7
      undefined
    console.log node_modules/next/dist/export/index.js:1
      > using build directory: /Users/zephraph/Code/oss/next-mdx-enhanced/__tests__/fixtures/layout-exports-ssg/.next
    console.log node_modules/next/dist/export/index.js:1
        copying "static build" directory
    console.log node_modules/next/dist/export/index.js:12
      > No "exportPathMap" found in "next.config.js". Generating map from "./pages"
    console.log node_modules/next/dist/export/index.js:1
        launching 7 workers
    console.log node_modules/next/dist/build/spinner.js:3
      Exporting (0/2)
    console.log node_modules/next/dist/export/index.js:1
      Exporting (1/2)
    console.log node_modules/next/dist/export/index.js:1
      Exporting (2/2)
    console.log node_modules/next/dist/export/index.js:1
      

  ● exports ssg functions from layout

    ENOENT: no such file or directory, open '/Users/zephraph/Code/oss/next-mdx-enhanced/__tests__/fixtures/layout-exports-ssg/out
/docs/intro.html'

      121 | 
      122 | function expectContentMatch(outPath, filePath, matcher) {
    > 123 |   const content = fs.readFileSync(path.join(outPath, filePath), 'utf8')
          |                      ^
      124 |   return expect(content).toMatch(matcher)
      125 | }
      126 | 

      at expectContentMatch (__tests__/integration.js:123:22)
      at Object.<anonymous> (__tests__/integration.js:44:3)
```

I can confirm that there's nothing in docs

Second run (pass):

```
 PASS  __tests__/integration.js
  ● Console

    console.log node_modules/next/dist/build/spinner.js:3
      Creating an optimized production build...
    console.log node_modules/next/dist/build/index.js:10
      undefined
    console.log node_modules/next/dist/build/index.js:12
      Compiled successfully.
      
    console.log node_modules/next/dist/build/spinner.js:3
      Automatically optimizing pages...
    console.log node_modules/next/dist/build/index.js:44
      undefined
    console.log node_modules/next/dist/build/utils.js:7
      Page                                                           Size     First Load
      ─ ● /docs/intro                                                1.49 kB     54.1 kB
      + shared by all                                                52.6 kB
        ├ static/pages/_app.js                                       597 B
        ├ chunks/611306fa591f4ba315cedb70a4862505b6b39dfc.d90c6a.js  6.6 kB
        ├ chunks/framework.7dc243.js                                 40 kB
        ├ runtime/main.dff265.js                                     4.71 kB
        └ runtime/webpack.b65cab.js                                  746 B
    console.log node_modules/next/dist/build/utils.js:7
      undefined
    console.log node_modules/next/dist/build/utils.js:7
      λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
      ○  (Static)  automatically rendered as static HTML (uses no initial props)
      ●  (SSG)     automatically generated as static HTML + JSON (uses getStaticProps)
    console.log node_modules/next/dist/build/utils.js:7
      undefined
    console.log node_modules/next/dist/export/index.js:1
      > using build directory: /Users/zephraph/Code/oss/next-mdx-enhanced/__tests__/fixtures/layout-exports-ssg/.next
    console.log node_modules/next/dist/export/index.js:1
        copying "static build" directory
    console.log node_modules/next/dist/export/index.js:12
      > No "exportPathMap" found in "next.config.js". Generating map from "./pages"
    console.log node_modules/next/dist/export/index.js:1
        launching 7 workers
    console.log node_modules/next/dist/build/spinner.js:3
      Exporting (0/2)
    console.log node_modules/next/dist/export/index.js:1
      Exporting (1/2)
    console.log node_modules/next/dist/export/index.js:1
      Exporting (2/2)
    console.log node_modules/next/dist/export/index.js:1
```

The file _is_ there this time 
image

**Comment by zephraph** [comment]
Okay, I have a rough hypothesis. 

I think that fs events haven't necessarily finished by the time the `compileNextjs` promise resolves.

**Comment by zephraph** [comment]
Okay, so, I updated how the tests run which resolves this weird issue.

Essentially instead of relying on the internal `build` and `export` methods, I'm just directly calling `next build` and `next export` via subprocesses. 

This _works_, but it doesn't help me understand _why_ this process is breaking down. For whatever reason when `build` + `export` are called in the same node process, not all of the files that are needed to be exported are actually exported. I don't know if this is because it's interpreting the build environment as something special or what's happening...

It also only seems to be related to this new SSG functionality. I'd almost be willing to be that it's a _really_ weird corner case. Maybe something to do with the [workers](https://github.com/zeit/next.js/blob/canary/packages/next/export/worker.js)?

Idk. @jescalan, let me know if this change is okay. It does slow down the tests quite a bit. Maybe that's acceptable given that there aren't many.

**Comment by zephraph** [comment]
Hey @jescalan, have you had a chance to check it out?

**Comment by jescalan** [comment]
Hi! I'm so sorry I missed all of these comments somehow. I will get back to you on this shortly :grin:

**Comment by zephraph** [comment]
Sorry for the repeated pings here 🙏.

If there's any changes I can make to help this along please let me know.

File: __tests__/integration.js [hunk 1]
```diff
@@ -4,6 +4,7 @@ const rmfr = require('rmfr')
 const nextBuild = require('next/dist/build').default
 const nextExport = require('next/dist/export').default
```
Comment (jescalan) [inline __tests__/integration.js:5]: I think we can drop these two lines now right?

Comment (zephraph) [reply __tests__/integration.js:5]: Oh, yep!

**Comment by jescalan** [comment]
Ok sorry for the unacceptably long delay, this is getting merged this week!

File: __tests__/integration.js (hunk 1 above)
Comment (zephraph) [reply __tests__/integration.js:5]: Actually, no, these are still used by `compileNextjsWithMockFunction`. That's still used by the `options.onContent` test.

**Comment by zephraph** [comment]
@jescalan I'm pretty sure this is good to go. Let me know if you see anything else and I'll try to respond asap.

**Comment by jescalan** [comment]
@zephraph one more quick as - could you rebase this? It says there are merge conflicts right now and I can't merge 😭

**Comment by zephraph** [comment]
Okay @jescalan, rebased.

**Comment by jescalan** [comment]
🙌🏼 shipping this out now as a minor version boost. thank you again for the contribution and I am really sorry it took me so long to follow through here. never again!
